src/launchpads/MK3/LaunchpadMK3.ts

```diff
@@ -0,0 +1,242 @@
+    }
+
+    const [r, g, b] = color;
+    const buttonMapped = this.mapButtonFromXy(button);
+
+    this.sendSysEx(...SysEx.setButtonColors(ButtonColor.rgb(
+      buttonMapped, r, g, b)));
+  }
+
+  /**
+   * @inheritDoc
+   */
+  flash(button: number|number[], color: number, colorB: number = 0): void {
+    const buttonMapped = this.mapButtonFromXy(button);
+
+    this.sendSysEx(...SysEx.setButtonColors(ButtonColor.flash(buttonMapped, color, colorB)));
+  }
+
+  /**
+   * @inheritDoc
+   */
+  pulse(button: number|number[], color: number): void {
+    const buttonMapped = this.mapButtonFromXy(button);
+
+    this.sendSysEx(...SysEx.setButtonColors(ButtonColor.pulse(buttonMapped, color)));
+  }
+
+  /**
+   * @inheritDoc
+   */
+  allOff(): void {
+    this.sendSysEx(14, 0);
+  }
+
+  /**
+   * @inheritDoc
+   */
+  closePorts(): void {
+    this.logDebug('Closing ports');
+
+    this.allOff();
+    this.input.closePort();
+    this.output.closePort();
+  }
+
+  /**
+   * @inheritDoc
+   */
+  eventNames(): string[] {
+    return [
+      'ready',
+      'rawMessage',
+      'buttonDown',
+      'buttonUp',
+    ];
+  }
+
+  /**
+   * @inheritDoc
+   */
+  parseButtonToXy(state: number, note: number): number[]|number {
+    if (!this.options.xyMode) {
+      return note;
+    }
+
+    const row = Math.floor(note / 10);
+    const col = (note - 1) % 10;
+
+    return [col, 9 - row];
+  }
+
+  /**
+   * @inheritDoc
+   */
+  mapButtonFromXy(xy: number[]|number): number {
+    const [x, y] = Array.isArray(xy) ? xy : [xy, 0];
+
+    if (!this.options.xyMode) {
+      return x;
+    }
+
+    return (9 - y) * 10 + x + 1;
+  }
+
+  private setupMessageHandler(): void {
+    this.input.on('message', (deltaTime: number, message: number[]) => {
+      this.logDebug(`m: ${message} d: ${deltaTime}`);
+      this.internalMessageHandler(message);
+    });
+  }
+
+  private internalMessageHandler(message: number[]): void {
+    this.emit('rawMessage', message);
+
+    const [state, note, value] = message;
+    const button = this.parseButtonToXy(state, note);
+
+    const upDown = Boolean(value) ? 'Down' : 'Up';
+    this.emit(`button${upDown}`, button, value);
+  }
+
+  private logDebug(...message: any[]): void {
+    if (this.options.debug) {
+      console.log('[Launchpad Debug]', ...message);
+    }
+  }
+}
+
+/**
+ * Specify the color of a single button
+ */
+export class ButtonColor {
+  public static staticColor(button: number, color: number) {
+    return [0, button, color];
+  }
+
+  public static flash(button: number, colorA: number, colorB: number) {
+    return [1, button, colorB, colorA];
+  }
+
+  public static pulse(button: number, color: number) {
+    return [2, button, color];
+  }
+
+  public static rgb(button: number, r: number, g: number, b: number) {
+    // r, g, b in 0..127
+    return [3, button, r, g, b];
+  }
+}
+
+/**
+ * Class to format SysEx messages into an array of numbers
+ */
+class SysEx {
```

Comment:
Could you put this in its own file?

Reply:
I'm not sure the message numbers are the same. I'll check it out.